File: src/io/github/hulang1024/chinesechess/scene/chessplay/rule/chess/ChessS.java
package io.github.hulang1024.chinesechess.scene.chessplay.rule.chess;

import io.github.hulang1024.chinesechess.scene.chessplay.rule.ChessPosition;
import io.github.hulang1024.chinesechess.scene.chessplay.rule.RoundGame;

/**
 * 兵
 * @author Hu Lang
 */
public class ChessS extends AbstractChess {
    @Override
    public boolean canGoTo(ChessPosition destPos, RoundGame game) {
        // 是否向前单步
        boolean isForward = game.isHostAtChessboardTop(host)
            ? pos.col + 1 == destPos.col
            : pos.col - 1 == destPos.col;

        // 判断是否过了河
        if (MoveRules.isInBoundary(game, host, destPos)) {
            // 过河之后既可以向前单步，也可以左或右移单步
            return isForward || Math.abs(destPos.col - pos.col) == 1;
        } else {
            // 过河之前只可以向前单步
            return isForward;
        }
    }
}

File: web-client/chinese-chess/src/scene/play/rule/chess/ChessS.ts
import ChessHost from "../chess_host";
import ChessPos from "../ChessPos";
import RoundGame from "../RoundGame";
import AbstractChess from "./AbstractChess";
import { isInBoundary } from "./move_rules";

/**
 * 兵
 */
export default class ChessS extends AbstractChess {
    constructor(pos: ChessPos, host: ChessHost) {
        super(pos, host);
    }

    canGoTo(destPos: ChessPos, game: RoundGame) {
        const rowOffset = destPos.row - this.pos.row;
        const colOffset = Math.abs(destPos.col - this.pos.col);

        // 是否向前单步
        let isForward = colOffset == 0
            && (game.isHostAtChessboardTop(this.host)
            ? rowOffset == +1
            : rowOffset == -1);

        // 判断是否过了河
        if (isInBoundary(game, this.host, destPos)) {
            // 过河之前只可以向前单步
            return isForward;
        } else {
            // 过河之后既可以向前单步，也可以左或右移单步
            return isForward || (Math.abs(rowOffset) == 0 && colOffset == 1);
        }
    }
}
